### PR TITLE
test(e2e): add e2e tests for  Date Picker component

### PR DESCRIPTION
**Pull request by nikvalor**
#### Please provide a link to the associated issue.
closes#[added-date-picker-core-tests](https://app.zenhub.com/workspaces/fundamental-library-5f32933d64202b0012e3572e/issues/sap/fundamental-ngx/4810)
#### Please provide a brief summary of this pull request.
Added date picker core e2e tests

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/main/CONTRIBUTING.md
- [x] tests for the changes that have been done
- [x] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist
- [x] Run npm run build-pack-library and test in external application

Documentation checklist:
- [x] update `README.md`
- [x] [Breaking Changes Wiki](https://github.com/SAP/fundamental-ngx/wiki/Breaking-Changes)
- [x] Documentation Examples
- [x] Stackblitz works for all examples

### CODE REVIEW

**Comment by nikvalor** [comment]
Project: null Platform A 
 Url: null 
 This link will expire at

**Comment by nikvalor** [comment]
Project: null Platform B
 Url: null 
 This link will expire at

File: e2e/wdio/platform/tests/time-picker.e2e-spec.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    })
+
+    describe('Check visual regression', function() {
+        it('should check examples visual regression', () => {
+            timePickerPage.saveExampleBaselineScreenshot('time-picker');
+            expect(timePickerPage.compareWithBaseline('time-picker')).toBeLessThan(1);
+        });
+
+        it('should check examples visual regression', () => {
+            scrollIntoView(activeTimePickerButton, 0);
```
Comment (Seamoo13) [inline e2e/wdio/platform/tests/time-picker.e2e-spec.ts:110]: do you need index 0? should pick 0 by default

**Review by Seamoo13 — APPROVED** [review]

**Comment by nikvalor** [comment]
Project: null Platform B
 Url: null 
 This link will expire at

**Comment by nikvalor** [comment]
Project: null Platform A 
 Url: null 
 This link will expire at

File: e2e/wdio/core/tests/date-picker.e2e-spec.ts [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+    afterEach(() => {
+        refreshPage();
+        waitForPresent(inputDatePicker);
+    }, 1);
+
+    it('Verify in all the form factor user is able to see the date picker button and input field ', () => {
+        const buttonsLength = getElementArrayLength(buttonDatePicker);
+        const inputsLength = getElementArrayLength(inputDatePicker);
+        expect(buttonsLength).toEqual(inputsLength);
+        for (let i = 1; i < buttonsLength; i++) {
```
Comment (AntonOlkhovskyi) [inline e2e/wdio/core/tests/date-picker.e2e-spec.ts:86]: `  waitForElDisplayed(buttonDatePicker, i);
  waitForElDisplayed(inputDatePicker, i);`

File: e2e/wdio/core/tests/date-picker.e2e-spec.ts [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+
+    afterEach(() => {
+        refreshPage();
+        waitForPresent(inputDatePicker);
+    }, 1);
+
+    it('Verify in all the form factor user is able to see the date picker button and input field ', () => {
+        const buttonsLength = getElementArrayLength(buttonDatePicker);
+        const inputsLength = getElementArrayLength(inputDatePicker);
+        expect(buttonsLength).toEqual(inputsLength);
```
Comment (AntonOlkhovskyi) [inline e2e/wdio/core/tests/date-picker.e2e-spec.ts:85]: Can you move expect to the end of the it. Looks more readable like this.

File: e2e/wdio/core/tests/date-picker.e2e-spec.ts [hunk 4]
```diff
@@ -0,0 +1,447 @@
...
+        click(yearInCalendarByValue(year2021));
+        expect(getText(buttonSelectYear)).toBe(year2021.toString());
+    });
+
+    it('Verify by default today date is focused date-picker', () => {
+        const activeButtonsLength = getElementArrayLength(activeButtonDatePicker);
+        for (let i = 0; i < activeButtonsLength; i++) {
+            if (i === 8) {
+                continue;
+            } else {
```
Comment (AntonOlkhovskyi) [inline e2e/wdio/core/tests/date-picker.e2e-spec.ts:119]: no need else here

File: e2e/wdio/core/tests/date-picker.e2e-spec.ts [hunk 5]
```diff
@@ -0,0 +1,447 @@
...
+
+    it('verify disabled date-picker pickers', () => {
+        const disabledDivDatePickerLength = getElementArrayLength(disabledDivDatePicker);
+        for (let i = 0; i < disabledDivDatePickerLength; i++) {
+            scrollIntoView(disabledDivDatePicker, i);
+            expect(getAttributeByName(disabledDivDatePicker, 'class')).toContain('is-disabled');
+        }
+    });
+
+    describe('Check visual regression', function() {
```
Comment (AntonOlkhovskyi) [inline e2e/wdio/core/tests/date-picker.e2e-spec.ts:313]: If you have this kind of repetitive check for each button you can use something similar to 

`    function checkElementStates(selector: string, tag: string, elementName: string, index: number = 0): void {
        checkHoverState(selector, tag + '-hover', elementName, index);
        checkFocusState(selector, tag + '-focus', elementName, index);
        checkActiveState(selector, tag + '-active', elementName, index);
    }`

to have one it for each button state + less tags name needs to be hardcoded in the tags fixture.

P.S. This is an example from select component for core.

File: e2e/wdio/core/tests/date-picker.e2e-spec.ts [hunk 6]
```diff
@@ -0,0 +1,447 @@
...
+
+        // skipped for now due to the issue with selected date for disabled components
```
Comment (AntonOlkhovskyi) [inline e2e/wdio/core/tests/date-picker.e2e-spec.ts:315]: Could you add link to the issue please.

**Review by AntonOlkhovskyi — CHANGES_REQUESTED** [review]

File: e2e/wdio/core/tests/date-picker.e2e-spec.ts [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+            click(activeButtonDatePicker, i);
+        }
+    });
+
+    it('verify selected date is showing in blue background', () => {
+        click(activeButtonDatePicker);
+        click(dayInCalendarButtonByValue('1'));
+        expect(getAttributeByName(activeInputDatePicker, 'ng-reflect-model')).toEqual(date);
+        click(activeButtonDatePicker);
+        expect(highlightedColor)
```
Comment (droshev) [inline e2e/wdio/core/tests/date-picker.e2e-spec.ts:254]: testing against hardcoded color should not be done

File: wdio-local.conf.js [hunk 8]
```diff
@@ -95,9 +95,9 @@ exports.config = {
             platformName: 'macOS 10.15',
             browserVersion: 'latest',
             acceptInsecureCerts: true,
-            'goog:chromeOptions': {
-                args: ['--window-size=1920,1417', '--headless']
-            },
+            // 'goog:chromeOptions': {
```
Comment (droshev) [inline wdio-local.conf.js:98]: can you delete it?

File: e2e/wdio/core/fixtures/appData/date-picker-contents.ts [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+export const internationalizationGerman = `1.${currentMonth}.2021`;
+export const internationalizationBulgarian = `1.${currentMonthWithZero}.2021 г.`;
+export const firstFormatterDate = `${currentMonthWithZero}/01/21`;
+export const secondFormatterDate = `${currentMonthWithZero}/01/21 to ${currentMonthWithZero}/15/21`;
+export const reactiveFormRoDDate = '10/1/2020 - 10/15/2020';
+export const reactiveFormDisableDate = `${currentMonth}/${new Date().getDate().toString()}/2021`;
+export const reactiveFormDisableRoDDate = `${currentMonth}/${new Date().getDate().toString()}/2021 - ${currentMonth}/${new Date().getDate().toString()}/2021`;
+export const date = `${currentMonth}/1/2021`;
+export const text = 'test';
+export const highlightedColor = ['rgb(8,84,160)', 'rgba(8,84,160,1)'];
```
Comment (droshev) [inline e2e/wdio/core/fixtures/appData/date-picker-contents.ts:28]: hardcoded color values

**Review by droshev — DISMISSED** [review]
see the comments

File: wdio-local.conf.js (hunk 8 above)
Comment (nikvalor) [reply wdio-local.conf.js:98]: Fixed

File: e2e/wdio/core/tests/date-picker.e2e-spec.ts (hunk 7 above)
Comment (nikvalor) [reply e2e/wdio/core/tests/date-picker.e2e-spec.ts:254]: We'll discuss this point on our meeting on Thursday and and will fix it if necessary after it

File: e2e/wdio/core/fixtures/appData/date-picker-contents.ts (hunk 9 above)
Comment (nikvalor) [reply e2e/wdio/core/fixtures/appData/date-picker-contents.ts:28]: https://github.com/SAP/fundamental-ngx/pull/4935#discussion_r607775724

File: e2e/wdio/core/tests/date-picker.e2e-spec.ts [hunk 10]
```diff
@@ -0,0 +1,387 @@
...
+            expect(getAttributeByName(disabledDivDatePicker, 'class')).toContain('is-disabled');
+        }
+    });
+
+    describe('Check visual regression', function() {
+
+        // skipped for now due to the issue with selected date for disabled components
+        xit('should check examples visual regression', () => {
+            datePickerPage.saveExampleBaselineScreenshot();
+            expect(datePickerPage.compareWithBaseline()).toBeLessThan(1);
```
Comment (Seamoo13) [inline e2e/wdio/core/tests/date-picker.e2e-spec.ts:317]: please change from 1 to 5

File: e2e/wdio/core/tests/date-picker.e2e-spec.ts [hunk 11]
```diff
@@ -0,0 +1,387 @@
...
+            scrollIntoView(buttonBulgarian);
+            checkElementStates(buttonBulgarian, bulgarianButton, button);
+        });
+    });
+
+    function checkElementHoverState(selector: string, tag: string, elementName: string, index: number = 0): void {
+        mouseHoverElement(selector, index);
+        saveElementScreenshot(selector, tag + getImageTagBrowserPlatform(), datePickerPage.getScreenshotFolder(), index);
+        expect(checkElementScreenshot(selector, tag + getImageTagBrowserPlatform(), datePickerPage.getScreenshotFolder(), index))
+            .toBeLessThan(2, `${elementName} element hover state mismatch`);
```
Comment (Seamoo13) [inline e2e/wdio/core/tests/date-picker.e2e-spec.ts:363]: please change from 2 to 5, same for below

**Review by Seamoo13 — APPROVED** [review]
approved, but please address small comment

**Review by AntonOlkhovskyi — APPROVED** [review]

**Review by InnaAtanasova — APPROVED** [review]

File: e2e/wdio/core/fixtures/appData/date-picker-contents.ts (hunk 9 above)
Comment (droshev) [reply e2e/wdio/core/fixtures/appData/date-picker-contents.ts:28]: Can you delete this?

**Review by droshev — COMMENTED** [review]
Can you delete the highlightColor var?

File: e2e/wdio/core/fixtures/appData/date-picker-contents.ts [hunk 12]
```diff
@@ -0,0 +1,31 @@
+export const currentMonth = new Date().getMonth() + 1;
+export const currentYear = new Date().getFullYear().toString();
+export const currentMonthWithZero = ('0' + `${currentMonth}`).slice(-2);
```
Comment (aberikashvili) [inline e2e/wdio/core/fixtures/appData/date-picker-contents.ts:3]: Concatenation could be done like this here: `
```
`0${currentMonth}`
```

**Pull request closed without merging**